sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java

```diff
@@ -2363,6 +2367,136 @@ public int hashCode() {
         return getClass().hashCode();
       }
     }
+
+    @Test
+    @Category({
+      ValidatesRunner.class,
+      UsesStatefulParDo.class,
+      UsesRequiresTimeSortedInput.class,
+      UsesStrictTimerOrdering.class
+    })
+    public void testRequiresTimeSortedInput() {
+      // generate list long enough to rule out random shuffle in sorted order
+      int numElements = 1000;
+      List<Long> eventStamps =
+          LongStream.range(0, numElements)
+              .mapToObj(i -> numElements - i)
+              .collect(Collectors.toList());
+      testTimeSortedInput(numElements, pipeline.apply(Create.of(eventStamps)));
+    }
+
+    @Test
+    @Category({
+      ValidatesRunner.class,
+      UsesStatefulParDo.class,
+      UsesRequiresTimeSortedInput.class,
+      UsesStrictTimerOrdering.class,
+      UsesTestStream.class
+    })
+    public void testRequiresTimeSortedInputWithTestStream() {
+      // generate list long enough to rule out random shuffle in sorted order
+      int numElements = 1000;
+      List<Long> eventStamps =
+          LongStream.range(0, numElements)
+              .mapToObj(i -> numElements - i)
+              .collect(Collectors.toList());
+      TestStream.Builder<Long> stream = TestStream.create(VarLongCoder.of());
+      for (Long stamp : eventStamps) {
+        stream = stream.addElements(stamp);
+      }
+      testTimeSortedInput(numElements, pipeline.apply(stream.advanceWatermarkToInfinity()));
+    }
+
+    @Test
+    @Category({
+      ValidatesRunner.class,
+      UsesStatefulParDo.class,
+      UsesRequiresTimeSortedInput.class,
+      UsesStrictTimerOrdering.class,
+      UsesTestStream.class
+    })
+    public void testRequiresTimeSortedInputWithLateData() {
+      // generate list long enough to rule out random shuffle in sorted order
+      int numElements = 1000;
+      List<Long> eventStamps =
+          LongStream.range(0, numElements)
+              .mapToObj(i -> numElements - i)
+              .collect(Collectors.toList());
+      TestStream.Builder<Long> input = TestStream.create(VarLongCoder.of());
+      for (Long stamp : eventStamps) {
+        input = input.addElements(TimestampedValue.of(stamp, Instant.ofEpochMilli(stamp)));
+        if (stamp == 100) {
+          // advance watermark when we have 100 remaining elements
+          // all the rest are going to be late elements
+          input = input.advanceWatermarkTo(Instant.ofEpochMilli(stamp));
+        }
+      }
+      testTimeSortedInput(
+          numElements - 100,
+          numElements - 1,
+          pipeline.apply(input.advanceWatermarkToInfinity()),
+          // Only direct runner is guaranteed to correctly drop elements after watermark  coming
+          // from different (sub-)partitions of TestStream. Cannot reference DirectRunner by
+          // class to avoid cyclic dependencies.
+          pipeline.getOptions().getRunner().getSimpleName().equals("DirectRunner"));
```

Comment:
Other runners (flink to be specific) might handle watermark differently. Only direct runner synchronously advances watermark, so that what you mark as "late" in TestStream will really end up late. When running the same TestStream with flink, the watermark might get updated at different times and thus the data might not be late and the output is indeed different.

Reply:
Is this a problem of runner (Flink) implementing TestStream sub-optimally or that Beam can not have a full control over watermark behavior? I think it is latter.
From comment, I was not sure if it is workaround of not testing on all runners because it is broken or a workaround of assigning which runner this make sense to run on.

Reply:
Yes, it is the latter. Direct runner is somewhat special, because it is intentionally designed to be single-process and thus can have stricter guarantees. Another option would be to split (basically copy & paste) this test to two versions - one that would reside in tests specific for DirectRunner and the other would be validates runner test. I think that this is probably not worth the code duplication.

Reply:
If it is about code duplication around setting up TestStream, that can be reused via shared routine. I would actually prefer cleaner association of which test logic make sense for which test category and avoid "sniffing context" where it is running. It did confuse at least myself.

Reply:
I'm afraid, that code reuse between the two test modules would not be straightforward and would introduce much more counter intuitive code.
The only option would be to split the test into two parts:
 a) validate that the logic of the late data dropping is correct (that is, test the annotation itself) (this would have to be placed outside validates runner suites)
 b) validate that runners obey the annotation (validates runner)
There would be some code duplication probably, but it might be worth it. I will look into that.